### DISCUSSION TITLE
Integration OGC API Tiles and OGC API Maps

### DISCUSSION THREAD

**Discussion by hblitza** [mapproxy/mapproxy#890] [Ideas]
**Occasion**: Discussion on Mailing List: https://lists.osgeo.org/pipermail/mapproxy-dev/2023-November/000058.html

**Aims:**
- MapProxy shall be able to support new OGC API Standards (as input and output services)

# Ideas: 💡

## A Hybrid approach "ogc api mapproxy bridge":
 - Use pygeoapi as facade to serve OGC API Tiles and Maps
 - Therefore use a cached WMTS as Input in pygeoapi (WMTSFacadeProvider) / WMS as Input (WMSFacadeProvider)
 - Implement WMTS in pygeoapi is nessessary
 - provide an ready-to-use docker example for this hybrid approach

## B Concept of integrate the OGC API Standards into MapProxy
 - Fully implement OGC API Tiles and Maps in MapProxy (for Input Sources and output services)
 - make use of classes that are used in pygeoapi. Therefore some refactoring might be neccessary

### Work so far 💪

- WMTSProvider for pygeoapi https://github.com/geopython/pygeoapi/pull/1473  by @simonseyock 
- Example configuration for using MapProxy with pygeoapi https://github.com/simonseyock/mapproxy-pygeoapi-example by @simonseyock 

## TODO 👷 

- Discussion of ideas
- Test Approach A and collect feedback
- Concrete steps for approach B (what classes can be used, what needs to be refactored) --> optimize synergy effects from both projects
- Sponsoring! Very important features to make MapProxy future-proof. All users will benefit from that.

**Comment by thorsten-reitz** [comment]
@hblitza @simonseyock We would be interested in supporting the development of OGC API, Maps in MapProxy, ideally to use it as a facade for existing WMS backend services. As this ticket states that sponsoring would be desired: How much would it take to sponsor this kind of development?

**Comment by weskamm** [comment]
Target
------

MapProxy shall support OGC API Maps and Tiles as input (mapproxy "source") and output (mapproxy "layer" or "service" / ogc api "collection").

The input could be implemented by writing mappers which convert between exisiting parts in mapproxy, e.g. 
  * WMS <-> OGC API Maps and 
  * WMTS <-> OGC API Tiles.

As the OGC API requires its own browsable landing page for the offered layers or `collections`, we should think about a way to merge the already exisiting mapproxy demo page with this new HTML API.

Where applicable the validations of the implementation should pass via: https://cite.ogc.org/teamengine/.
Currently the support for OGC API Maps is missing there.


After a closer look on the specification, we decided to flag the subjects in three categories: "Needs to be implemented", "Not needed" and "optional".
This is the current roadmap for the implementation of the OGC API, which still is subject to change and is open for discussion.
We listed the relevant OGC Specification and their requirement classes:

Common - Part 1
---------------

* Core Requirements Class
    * HTTP 1.1
      * Needs to be implemented
    * HTTP Status Codes
      * This has to be checked. MapProxy should in general use the correct status codes, independent of the OGC API
    *  Query parameters
      * Needs to be implemented
    * Web Caching
      * Would be nice to have. Would need to be implemented in MapProxy in general. Can be done outside of OGC API Support.
    * Support for Cross-Origin Requests
      * Needs to be implemented
    * String Internationalization
      * Not needed. Everything is in english.
    * Resource Encodings
      * Needs to be implemented. HTML / JSON
    * Parameter Encoding
      * Needs to be implemented / should be respected in the OGC API Part of MapProxy
    
* Landing Page
    * API Landing page
        * Needs to be implemented
        * Links to collections, conformance classes, tileset matrixes, open api and should be included
        * should be available as html/json
        * HTML should include link to json version
    * API Definition
        Needs to be implemented -> Open API 3.0
    * Declaration of Conformance Classes
        Needs to be implemented

* Encoding Classes
    * HTML and JSON need to be implemented

* OpenAPI 3.0
    * optional but would be nice to have

* Building Blocks
    * Seems to be useful for major developments at the base of the API, should not be needed


Common - Part 2
---------------

* In general mapproxy should handle the term "collection" as the already exisiting "layers"
* The output of the `/collections` endpoint should list the available layers with their respective config

* Requirements Class “Collections”
    * List of Collections Resource (/collections)
      * Needs to be implemented
    * Collection Description Resource (/collections/{collectionId})
      * Needs to be implemented
      * Attribution: We could introduce attributions for layers/collection.
        * Not needed
      * Item type
       * Not applicable as we don't have items and use the collections as placeholders for layers
      * Links
        * Needs to be implemented
      * Extent
        * Needs to be implemented (either include as text or in a map)
      * Either this page directly includes a map viewer or should link to one.

* Requirements Class “Uniform Multi-Dimension Collection”
    * Not needed

* Requirements Classes for Encodings
    HTML, JSON need to be implemented

Maps
----
* Could convert requests / use existing WMS implementation under the hood.
  * should integrate cleanly in exisiting codebase
* should make use of mapproxys cache ability

Conformance Classes

* Requirements Class “Core"
    * Needs to be implemented

* Requirements Class “Map Tilesets”
    * Needs to be implemented

* Requirements Class “Background”
    * Needs to be implemented
    * May reuse / enhance existing code of source parameter `transparent_color` and handling of `transparent` param in WMS requests

* Requirements Class “Collection Selection”
    * Needs to be implemented

* Requirements Class “Scaling”
    * Parameters width and height
      * Needs to be implemented
    * Parameter scale-denominator
      * Optional, would be nice

* Requirements Class “Display Resolution"
    * Needs to be implemented
    * replaces `dpi`

* Requirements Class “Spatial Subsetting”
    * bbox: Needs to be implemented
    * subset: Not needed
    * center, width, height: Not needed

* Requirements Class “Date and Time”
    * Not needed

* Requirements Class “General Subsetting”
    * Not needed

* Requirements Class “Coordinate Reference System"
    * Needs to be implemented

* Requirements Class “Orientation” 
    * Optional

* Requirements Class “Custom Projection CRS”
    * Not needed

* Requirements Class “Collection Map” 
    * Needs to be implemented

* Requirements Class “Dataset Map”
    * Not needed (no dataset)

* Requirements Class “Styled Maps”
    * Not needed

* Requirements Classes for Encodings

  * PNG (https://www.opengis.net/spec/ogcapi-maps-1/1.0/req/png)
    * Needs to be implemented

  * JPEG (https://www.opengis.net/spec/ogcapi-maps-1/1.0/req/jpeg)
    * Needs to be implemented

  * JPEG XL (https://www.opengis.net/spec/ogcapi-maps-1/1.0/req/jpegxl)
    * Not needed

  * TIFF (https://www.opengis.net/spec/ogcapi-maps-1/1.0/req/tiff)
    * Not needed

  * SVG (https://www.opengis.net/spec/ogcapi-maps-1/1.0/req/svg)
    * Not needed

  * HTML (https://www.opengis.net/spec/ogcapi-maps-1/1.0/req/html)
    * Needs to be implemented

* Requirements Class “API Operations”
    * Needs to be implemented -> as OpenAPI 3.0

* Requirements Class “CORS”
    * Needs to be implemented


Tiles
-----
* Could convert requests / use existing WMTS implementation under the hood.
  * should integrate cleanly in exisiting codebase
* Should make use of mapproxys cache ability

* Requirements Class “CORE”
  * A Tile
    * Base idea of tiling mechanism and retrieval via templated URI (tilematrixset/tilcol/tilerow)
    * we think that all of this are more or less mandatory and should be supported
* Requirements Class “TILESET"
  * Example tileset: https://demo.ldproxy.net/daraa/tiles?f=json
  * Needs to be implemented
* Requirements Class “TILESET LIST"
  * Needs to be implemented

* Requirements Class “DATASET TILESETS"
  * Seems like it could be useful / needed in conjunction with multimapproxy, where every mapproxy represents a single "rootResource" a.k.a dataset.
  * Implementation may not be necessary
* Requirements Class “GEODATA TILESETS"
  * One further step in the chain
  * Needs to be implemented
* Requirements Class “COLLECTIONS SELECTION"
  * acts as a filter of "GEODATA TILESETS"
  * When not implementing the idea of (multiple) "dataset", this may also be unnecessary to implement
* Requirements Class “DATETIME"
  * Not needed
* Requirements Class “OPENAPI SPEC 3.0
  * Needs to be implemented -> as OpenAPI 3.0
* Requirements Class “XML TILESET METADATA"
  * Not needed
* Requirements Class “TILE ENCODING"
  * PNG (https://docs.ogc.org/is/20-057/20-057.html#rc_png)
    * Needs to be implemented
  * JPEG (https://docs.ogc.org/is/20-057/20-057.html#rc_jpeg)
    * Needs to be implemented
  * TIFF
    * optional
  * NetCDF
    * optional
  * GeoJSON
    * not needed
  * Mapbox Vector Tiles
    * not needed


OGC Two Dimensional Tile Matrix Set and Tile Set Metadata
-----------------------------------------------------------

  * TileMatrixSet: 
    * Needs to be implemented
  * VariableMatrixWidth
    * Optional
  * TileMatrixSetLimits
    * May be useful, optional
  * TileSetMetadata
    * Needs to be implemented
  * The XML Requirements are not needed
  * The JSON Requirements should be implemented

**Comment by weskamm** [comment]
Big thanks to @rouault for adding support via https://github.com/mapproxy/mapproxy/pull/1268 and https://github.com/mapproxy/mapproxy/pull/1284

This can be now be closed :partying_face:
